bite-size-notes/week-5-e-SCvx-and-GuSTO-SCP.md

```diff
@@ -0,0 +1,127 @@
+
+## SCOPE
+This note is all about two specific **Sequential Convex Programming (SCP)** algorithms; **Successive Convexification (SCvx)** and **Guaranteed Sequential Trajectory Optimization (GuSTO)**. We will briefly introduce SCP before diving into the two algorithms discussing what 
+they are used for and how to use them. This is a collection of notes taken from a paper written by researchers from the University of 
+Washington's Autonomous Controls Laboratory (UW ACL), the developers of the two algorithms[1].
+
+## Motivation and Introduction
+Generating safe and reliable trajectories is crucial in the autonomous world through the computation of multidimensional discrete state and control signals within a set of constraints that satisfies a set of specifications while optimizing for a mission objective. This challenges engineers and researchers to develop algorithms that focus on the safety, performance, and trustworthiness of trajectory generation. Numerical Optimization presents a great solution for the generation and optimization of trajectories to meet objectives so it can be expressed as an Optimal Control Problem.
+
+The optimization problem is solved through the following:
+> 1. **Formulation:** specifications of how the functions to be minimized and constraints to be satisfied are expressed mathematically.
+> 2. **Discretization:** approximation of the infinite-dimensional state and control signals by a finite-dimensional set of basis functions.
+> 3. **Numerical Optimization:** iterative computation of an optimal solution of the discretized problem.
+
+In terms of efficiency and reliability for numerical optimization, discretized convex problems are the way to go, which is the main motivation for **Convex Programming** or **Convex Optimization**. However, you may find that most problems in the real world are **nonlinear (nonconvex)**, which leads to the topic of **Convexification** and **SCP** for trajectory generation of nonlinear dynamic systems based on convex optimization. The main idea of SCP is iterative convex approximation as the name suggests. The general outline of an SCP algorithm is shown in **Fig. 1** where it starts with an initial trajectory guess that goes through an iteration scheme that optimizes the trajectory until it converges into a feasible solution at which the iteration is stopped. With SCP, safe and reliable trajectories can be generated for nonconvex functions.
+
+![SCP Block Diagram](figs/SCP_diagram.png)
+**Figure 1:** Block diagram of a typical SCP algorithm acquired from [1].
+
+An algorithm hierarchy in **Fig. 2** shows where SCP lies. Typically, the convex solver in layer 2 of **Fig. 1** iteratively calls an **Interior Point Method (IPM)** algorithm from layer 3 in **Fig. 2** that solves a convex problem with linear equality and convex inequality constraints as a sequence of linear equality-constrained problems and thus, IPM iteratively calls the algorithm in layer 2 in **Fig. 2**.
```

Comment:
What is the layer 2 of Fig. 1? Is it the layer 2 of Fig. 2?

Reply:
The layers are denoted by the circled numbers. Layer 2 of SCP calls an IPM algorithm, the layer 3 of Fig. 2. Then the IPM algorithm calls layer 2 of Fig. 2.